lib/dynamoid/criteria/chain.rb

```diff
@@ -176,6 +176,27 @@ def count
         end
       end
 
+      # Returns the first item matching the criteria.
+      #
+      #   Post.where(links_count: 2).first
+      #
+      # Applies `record_limit(1)` to ensure only a single record is fetched
+      # when no non-key conditions are present.
+      #
+      # If used without criteria it just returns the first item of some
+      # arbitrary order.
+      #
+      #   Post.first
+      #
+      # @return [Model|nil]
+      def first
+        if @key_fields_detector.non_key_present?
```

Comment:
I believe this condition should be more strict - only Query operations w/o non key conditions should be optimized.

Condition on a hash key should be equality like `{ id: 1 }`. Not-equality condition for a hash key means the Scan operation will be used instead of the Query. And Scan operations combined with `record_limit` method call leads to the issue with multiple requests.

Reply:
actually, already the case. Maybe naming of `KeyFieldsDetector#non_key_present?` in  confused you. But naming of `KeyFieldsDetector#key_present?` doesn't tell me that equality operation is checked use either. So, added a spec for it.

Reply:
Agree 👍